feat(ngcc): support `__spreadArray` helper as used by TypeScript 4.2

JoostK opened the pull request:
In TypeScript 4.2 the `__spread` and `__spreadArrays` helpers were both
replaced by the new helper function `__spreadArray` in
microsoft/TypeScript#41523. These helpers may be used in downleveled
JavaScript bundles that ngcc has to process, so ngcc has the ability to
statically detect these helpers and provide evaluation logic for them.
Because Angular is adopting support for TypeScript 4.2 it becomes
possible for libraries to be compiled by TypeScript 4.2 and thus ngcc
has to add support for the `__spreadArray` helper. The deprecated
`__spread` and `__spreadArrays` helpers are not affected by this change.

Closes #40394

JoostK committed:
feat(ngcc): support `__spreadArray` helper as used by TypeScript 4.2

In TypeScript 4.2 the `__spread` and `__spreadArrays` helpers were both
replaced by the new helper function `__spreadArray` in
microsoft/TypeScript#41523. These helpers may be used in downleveled
JavaScript bundles that ngcc has to process, so ngcc has the ability to
statically detect these helpers and provide evaluation logic for them.
Because Angular is adopting support for TypeScript 4.2 it becomes
possible for libraries to be compiled by TypeScript 4.2 and thus ngcc
has to add support for the `__spreadArray` helper. The deprecated
`__spread` and `__spreadArrays` helpers are not affected by this change.

Closes #40394

petebacondarwin reviewed (approved):
Ahead of the game @JoostK

petebacondarwin commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/ts_helpers.ts:
```diff
@@ -35,3 +35,27 @@ export class SpreadHelperFn extends KnownFn {
+      return DynamicValue.fromInvalidExpressionType(node, from);
+    }
+
+    return from.concat(to);
```
Oops, I spoke too soon. This should be

```suggestion
    return to.concat(from);
```

JoostK replied:
That is entirely correct! I didn't notice as the evaluator spec are part of `ngtsc/partial_evaluator` and I only ran the `ngcc` test targets.

crisbeto reviewed (commented):
I don't have a lot of context here, but do we also have to account for the `__read` helper? It seems to be used together with `__spreadArray`. See https://github.com/angular/angular/pull/41158/files#diff-e5db61a4fea6171227d1b25300c43bc2f33613fefb1d98bb37ad746264c3661bR3006

JoostK commented:
> I don't have a lot of context here, but do we also have to account for the `__read` helper? It seems to be used together with `__spreadArray`. See https://github.com/angular/angular/pull/41158/files#diff-e5db61a4fea6171227d1b25300c43bc2f33613fefb1d98bb37ad746264c3661bR3006

Very good point, yes we do! The `__read` helper is emitted if `downlevelIteration` is set, which is quite likely for libraries (ng-packagr has that on by default). I'll add another commit to add support for `__read`

JoostK committed:
fixup! feat(ngcc): support `__spreadArray` helper as used by TypeScript 4.2

JoostK committed:
feat(ngcc): support `__read` helper as used by TypeScript 4.2

This commit complements the support for the `__spreadArray` helper that
was added in microsoft/TypeScript#41523. The prior helpers `__spread`
and `__spreadArrays` used the `__read` helper internally, but the helper
is now emitted as an argument to `__spreadArray` so ngcc now needs to
support evaluating it statically. The real implementation of `__read`
reads an iterable into an array, but for ngcc's static evaluation
support it is sufficient to only deal with arrays as is. Additionally,
the optional `n` parameter is not supported as that is only emitted for
array destructuring syntax, which ngcc does not have to support.

petebacondarwin commented:
> but for ngcc's static evaluation support it is sufficient to only deal with arrays as is

Is it not possible for a developer to use an iterator in a spread expression?

petebacondarwin reviewed (approved):
I would like to know why the implementation of `__read` doesn't require iterator support. But that could be added on as a subsequent PR if it is found to be necessary.

JoostK commented:
> > but for ngcc's static evaluation support it is sufficient to only deal with arrays as is
> 
> Is it not possible for a developer to use an iterator in a spread expression?

AFAIK this would not be statically evaluable, i.e. instantiating a `Set` needs the `new` operator which is not supported, and generator functions nor the iterator protocol is implemented by ngtsc's evaluator.

AndrewKushnir commented:
[Presubmit](http://test/OCL:363198615:BASE:363198630:1615911529715:d3c1b3fe).
